modules/appgw/variables.tf

```diff
@@ -75,95 +105,515 @@ variable "capacity_max" {
+  - `ssl_certificate_path`                       - (`string`, optional) Path to the file with tThe base64-encoded PFX certificate data.
+  - `ssl_certificate_pass`                       - (`string`, optional) Password for the pfx file specified in data.
+  - `ssl_certificate_vault_id`                   - (`string`, optional) Secret Id of (base-64 encoded unencrypted pfx) Secret
+                                                   or Certificate object stored in Azure KeyVault.
+  - `custom_error_pages`                         - (`map`, optional) Map of string, where key is HTTP status code and value is
+                                                   error page URL of the application gateway customer error.
+  EOF
+  type = map(object({
+    name                     = string
+    port                     = number
+    protocol                 = optional(string, "Http")
+    host_names               = optional(list(string))
+    ssl_profile_name         = optional(string)
+    ssl_certificate_path     = optional(string)
+    ssl_certificate_pass     = optional(string)
+    ssl_certificate_vault_id = optional(string)
+    custom_error_pages       = optional(map(string), {})
+  }))
+  validation {
+    condition = alltrue(flatten([
+      for _, listener in var.listeners : [
+        contains(["Http", "Https"], listener.protocol)
+    ]]))
+    error_message = "Possible values for `protocol` are Http and Https."
+  }
+  validation {
+    condition = alltrue(flatten([
+      for _, listener in var.listeners : (listener.port >= 1 && listener.port <= 65535)
+    ]))
+    error_message = "The listener `port` should be a valid TCP port number from 1 to 65535."
+  }
+  validation {
+    condition = alltrue(flatten([
+      for _, listener in var.listeners : (listener.protocol == "Https" ? try(length(coalesce(listener.ssl_certificate_vault_id, listener.ssl_certificate_path)), -1) > 0 : true)
+    ]))
+    error_message = "If HTTPS protocol is used, then SSL certificate (from file or Azure Key Vault) is required"
+  }
+  validation {
+    condition = alltrue(flatten([
+      for _, listener in var.listeners : (listener.protocol == "Https" ? try(length(listener.ssl_certificate_pass), -1) >= 0 : true)
+    ]))
+    error_message = "If HTTPS protocol is used, then SSL certificate password is required"
+  }
+  validation {
+    condition     = length(flatten([for _, listener in var.listeners : listener.name])) == length(distinct(flatten([for _, listener in var.listeners : listener.name])))
+    error_message = "The `name` property has to be unique among all listeners."
+  }
+}
+
+variable "backend_pool" {
+  description = <<-EOF
+  Backend pool.
+
+  Object contains attributes:
+  - `name`         - (`string`, optional) name of the backend pool.
+  - `vmseries_ips` - (`list`, optional) IP addresses of VMSeries' interfaces that will serve as backends for the Application Gateway.
+  EOF
+  type = object({
```

Comment:
yes , because according to https://developer.hashicorp.com/terraform/language/expressions/type-constraints:

_Terraform typically returns an error when it does not receive a value for specified object attributes. When you mark an attribute as optional, Terraform instead inserts a default value for the missing attribute. This allows the receiving module to describe an appropriate fallback behavior._

so I no `backend_pool` is provided, following object is used:
```
    backend_address_pool {
        fqdns        = []
        id           = "***"
        ip_addresses = []
        name         = "vmseries"
    }
```